### PR TITLE
Fix a difference of address in Request structure at proxy environment

### LINKED ISSUE #60544: kubectl top is failed in proxy environment
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
 /kind bug
> /kind feature

**What happened**:
`kubectl top` command is failed by the following error message when running `kubectl top` in proxy environment(export http_proxy=...).
```
error: failed to unmarshall heapster response: invalid character '\x15' looking for beginning of value
```

This error was caused by accessing the server listening on https with http.
Actually, kube-apiserver was accessed by http request. (I found a following kube-apiserver log)
```
http: TLS handshake error from 10.130.117.156:51248: tls: first record does not look like a TLS handshake
```

`kubectl top` command accesses to cluster addon heapster.
heapster listens to access of http.
But, kube-apiserver listens to access of https.

I think that the reason is a difference of destination address between Request structure and URL structure (in Request structure).
* Request structure is in RoundTrip function in staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go

Request structure example is shown as follow.

* http://10.244.0.121:8082 = Heapster Pod URI
  https://10.130.116.37:6443 = kube-apiserver URI

```
Method:GET
URL:http://10.244.0.121:8082/apis/metrics/v1alpha1/nodes?labelSelector=
Proto:HTTP/2.0
ProtoMajor:2
ProtoMinor:0
Header:map[
	X-Forwarded-For:[10.130.116.37]
	Accept:[application/json, */*]
	User-Agent:[kubectl/v1.9.3 (linux/amd64) kubernetes/d283541]
	Accept-Encoding:[gzip]
]
Body:<nil>
GetBody:<nil>
ContentLength:0
TransferEncoding:[]
Close:false
Host:10.130.116.37:6443
Form:map[]
PostForm:map[]
MultipartForm:<nil>
Trailer:map[]
RemoteAddr:10.130.116.37:37458
RequestURI:/api/v1/namespaces/kube-system/services/http:heapster:/proxy/apis/metrics/v1alpha1/nodes?labelSelector=
TLS:0xc42bbe2420
Cancel:<nil>
Response:<nil>
ctx:0xc42bc39740}I0223 01:58:34.062603
```

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9.3
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): CentOS
- Kernel (e.g. `uname -a`):
- Install tools: kubeadm
- Others: 
  create kubernetes cluster in proxy environment
    export http_proxy=...
    export https_proxy=...

### PR DESCRIPTION
**Pull request by ghost**
**What this PR does / why we need it**:
Accessing to cluster addon pod fails in proxy environment.
\* cluster addon pod listens to access of http.
The reason is a difference of destination address between Request structure and URL structure (in Request structure).
\* Request structure is in RoundTrip function in staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go

This PR fixes ServeHTTP function (in staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go), because the function calls RoundTrip function by new Request structure.
\* RoundTrip function (in staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go) is point at issue.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #60544

**Special notes for your reviewer**:

**Release note**:

`NONE`

/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go [hunk 1]
```diff
@@ -223,6 +223,8 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
 	newReq.Header = utilnet.CloneHeader(req.Header)
 	if !h.UseRequestLocation {
 		newReq.URL = &loc
+		newReq.Host = loc.Host
+		newReq.RequestURI = loc.Path
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go]: this really needs a test demonstrating the issue this resolves and preventing regression

**Comment by shun-miyoshi-com** [comment]
@liggitt Thanks your review.
I'm sorry, I don't know what kind of test you want.
Do you want following test?
* Input:
    req *http.Request
      req.URL = \<address for service1\>
      req.Host and req.RequestURI = \<address for service2\>
* Expected:
    target address of req.URL and target address of req.Host, req.RequestURI is same.

**Comment by jdumars** [comment]
/ok-to-test
Hi, please sort out your release note label when you can.  Thanks!

**Comment by yliaog** [comment]
Please add a test case in https://github.com/kubernetes/kubernetes/blob/ae1fc13aee81e66b9b74a5fb881ff3f90463ff4e/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go#L149

**Comment by shun-miyoshi-com** [comment]
Hi, @yliaog
Umm...
I rewrited a local valiable `newReq` in ServeHTTP method.
I don't know the best way to observe `newReq` from outside (TestServeHTTP).

**Comment by yliaog** [comment]
newReq is passed to the backend by the proxy, you should be able to check it like this: https://github.com/kubernetes/kubernetes/blob/ae1fc13aee81e66b9b74a5fb881ff3f90463ff4e/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go#L300

**Comment by shun-miyoshi-com** [comment]
Hi, @yliaog
Is it true?
I wrote following test code, but the result was not good.
I rewrote newReq.RequestURI of upgradeware.go directly in order to test.
But req variable in ServeHTTP of upgradeaware_test.go was not changed.

+ upgradeware.go
``` go
if !h.UseRequestLocation {
  newReq.URL = &loc
  newReq.Host = loc.Host
  newReq.RequestURI = "test" //loc.Path
}
```

+ upgradeaware_test.go
``` go
func (s *SimpleBackendHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
    s.requestURL = *req.URL
    s.requestHeader = req.Header
    s.requestMethod = req.Method
    f, _ := os.OpenFile("result.txt", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
    defer f.Close()
    fmt.Fprintf(f, "%+v\n", req)
    ...
}
```

+ result.txt
``` txt
...
&{Method:GET URL:/some/path Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[X-Forwarded-For:[127.0.0.1] X-Forwarded-Uri:[/some/path] User-Agent:[Go-http-client/1.1] Accept-Encoding:[gzip]] Body:{} GetBody:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:127.0.0.1:38609 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr:127.0.0.1:42424 RequestURI:/some/path TLS:<nil> Cancel:<nil> Response:<nil> ctx:0xc4201b1bc0}
```

**Comment by shun-miyoshi-com** [comment]
Could anyone help me create a test?

**Pull request closed without merging**